worker/upload.go

```diff
@@ -0,0 +1,1154 @@
+
+			// return the first unused queue
+		search:
+			for _, q := range allowed {
+				for _, shardID := range history {
+					if _, ok := u.completed[j.uploadID][shardID][q.fcid]; !ok {
+						continue search
+					}
+				}
+				return q
+			}
+			fmt.Printf("DEBUG PJ: %v | %v | no queue yet for sector %d, overdrive %v, allowed %d waiting on %d shards to complete (%v)\n", j.uploadID, j.shardID, j.sectorIndex, j.overdrive, len(allowed), len(history), history)
+			return nil
+		}(); queue != nil {
+			return queue
+		}
+
+		// otherwise keep waiting
+		select {
+		case <-u.sectorCompletedTriggers[j.uploadID]:
+			continue loop
+		case <-j.requestCtx.Done():
+			break loop
+		}
+	}
+
+	return nil
+}
+
+func (u *uploader) update(contracts []api.ContractMetadata, bh uint64) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+
+	// build map
+	c2m := make(map[types.FileContractID]api.ContractMetadata)
+	for _, c := range contracts {
+		c2m[c.ID] = c
+	}
+
+	// recreate the pool
+	var i int
+	for _, q := range u.contracts {
+		if _, keep := c2m[q.fcid]; !keep {
+			continue
+		}
+		delete(c2m, q.fcid)
+		u.contracts[i] = q
+		i++
+	}
+	for j := i; j < len(u.contracts); j++ {
+		u.contracts[j] = nil
+	}
+	u.contracts = u.contracts[:i]
+
+	// add missing uploaders
+	for _, contract := range c2m {
+		queue := u.newQueue(contract)
+		u.contracts = append(u.contracts, queue)
+		go processQueue(u.hp, u.rl, queue)
+	}
+
+	// update queue blockheight
+	for _, q := range u.contracts {
+		q.updateBlockHeight(bh)
+	}
+}
+
+func processQueue(hp hostProvider, rl revisionLocker, q *uploadQueue) {
+outer:
+	for {
+		// wait for work
+		select {
+		case <-q.queueChan:
+		case <-q.stopChan:
+			return
+		}
+
+		for {
+			// check if we are stopped
+			select {
+			case <-q.stopChan:
+				return
+			default:
+			}
+
+			// pop the next job
+			job := q.pop()
+			if job == nil {
+				continue outer
+			}
+
+			// skip if job is done
+			if job.done() {
+				continue
+			}
+
+			// execute it
+			var root types.Hash256
+			start := time.Now()
+			err := rl.withRevision(job.requestCtx, defaultRevisionFetchTimeout, q.fcid, q.hk, q.siamuxAddr, lockingPriorityUpload, q.blockHeight(), func(rev types.FileContractRevision) error {
+				var err error
+				root, err = job.execute(hp, rev)
+				return err
+			})
+
+			// track the error, but only if the job is not a cancelled overdrive
+			canceledOverdrive := job.done() && job.overdrive && err != nil
+			if !canceledOverdrive {
+				q.track(err, time.Since(start))
+			}
+
+			if err != nil {
+				job.fail(err)
+			} else {
+				job.succeed(root)
+			}
+		}
+	}
+}
+
+func (q *uploadQueue) blockHeight() uint64 {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.bh
+}
+
+func (q *uploadQueue) estimate() float64 {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// fetch average speed
+	speed := q.statsSpeed.average()
+	if speed == 0 {
+		speed = math.MaxFloat64
+	}
+
+	data := (len(q.queue) + 1) * rhpv2.SectorSize
+	return float64(data) * 0.000008 / speed
```

Comment:
If we use `bytes per millisecond` instead of mbs this math here also needs to be updated to get rid of the `0.000008`.